[clang] Skip auto-init on scalar vars that have a non-constant Init and no self-ref

jvoung opened the pull request:
In that scalar case, the Init should initialize the auto var before use.
The Init might use uninitialized memory from other sources (e.g., heap)
but auto-init did not help us in that case because the auto-init would
have been overwritten by the Init before use.

For non-scalars e.g., classes, the Init expr might be a ctor call that
leaves uninitialized members, so we leave the auto-init there.

The motivation is to have less IR for the optimizer to later remove, which
may not be until a fairly late pass (DSE) or may not get optimized in lower
optimization levels like O1 (no DSE) or sometimes due to derefinement.
This is ~10% less left-over auto-init in O1 in a few examples checked

jvoung committed:
Skip auto-init on scalar vars that have a non-constant Init and no self-ref

In that scalar case, the Init should initialize the auto var before use.
The Init might use uninitialized memory from other sources (e.g., heap)
but auto-init did not help us in that case because the auto-init would
have been overwritten by the Init before use.

For non-scalars e.g., classes, the Init expr might be a ctor call that
leaves uninitialized members, so we leave the auto-init there.

The motivation is to have less IR for the optimizer to reduce, which
may be in a fairly late pass (DSE) or may not get optimized in lower
optimization levels like O1 (no DSe). This is ~10% less left-over
auto-init in O1 in a few examples checked.

jvoung committed:
Clean up test checks (some blank CHECKs and a mislabeled check prefix)

ramosian-glider commented:
Drive-by comment: can you please either merge the two commits together, or change their descriptions so that they are not 100% identical?

ramosian-glider commented on clang/test/CodeGenCXX/auto-var-init-max-size.cpp:
```diff
@@ -15,7 +15,7 @@ struct Foo {
 
 int foo(unsigned n) {
   bool var_size_1;
-  long var_size_8 = 123;
```
Can we have both the test cases in which scalar variables are auto-inited and where the initialization is skipped?
Something along the lines of:

```
long var_size_8_init = 123;
long var_size_8_noinit;
long var_size_8_init_later;
...
var_size_8_init_later = 456;
```

jvoung replied:
Sure! Added those cases.

ramosian-glider replied:
LGTM

jvoung committed:
Add a few more cases to auto-var-init-max-size test (init, noinit, later)

We check that the init doesn't have auto-init of pattern or 0
and we check that noinit and init_later do.

jvoung commented:
> Drive-by comment: can you please either merge the two commits together, or change their descriptions so that they are not 100% identical?

Oops, changed the later commit description!

haopliu commented on clang/test/CodeGenCXX/trivial-auto-var-init-skip-scalar-with-nonconst-init.cpp:
```diff
@@ -0,0 +1,110 @@
+
+// Scalar with a self-reference: does need auto-init.
+// UNINIT-LABEL:  test_selfinit_lambda_call(
+// ZERO-LABEL:    test_selfinit_lambda_call(
```
Nit, if "UNINIT-LABEL", "ZERO-LABEL", and "PATTERN-LABEL" are always the same, just define a common label like "CHECK-LABEL:" to avoid redundancy?

jvoung replied:
I'm not sure if a common "CHECK-LABEL:" would have the same semantics and help split the stream for matching when there is a different prefix like "ZERO: " (or "PATTERN:") that follows, or if the CHECK and the ZERO end up as a independent scans. If it's an independent scan, then "CHECK-LABEL" wouldn't have the same semantics as "ZERO-LABEL:" (which helps split the stream for the "ZERO:" checks that follow -- e.g., in case there a multiple places that could match something generic like "store i32 0, ", the preceding ZERO-LABEL help disambiguate which to match).

I'd prefer to keep the redundancy just to be safe and more clear that the X-LABEL are for the X that it's associated with, unless you're pretty sure it can cross over different check prefixes.

haopliu commented:
LGTM, thanks!

ilya-biryukov commented on clang/lib/CodeGen/CGDecl.cpp:
```diff
@@ -1972,7 +1972,20 @@ void CodeGenFunction::EmitAutoVarInit(const AutoVarEmission &emission) {
+      // It may be that the Init expression uses other uninitialized memory,
+      // but auto-var-init here would not help, as auto-init would get
+      // overwritten by Init.
+      if (!D.getType()->isScalarType() || isAccessedBy(D, Init)) {
```
Could you check what happens with `-fblocks` that capture the variable if they're part of its initializer?
There is some special-case handling for this (when `capturedByInit == true`), I feel we might not be catching that case here.

See https://gcc.godbolt.org/z/cEexo5n77 for an example.
```cpp
void foo();
void test() {
 __block int calls = ^() {
    while (++calls < 100) {
        foo();
    }
    return calls;
 }();
}
 
```

jvoung replied:
Great question, thank you for reviewing!

I tried that example and it does appear to get auto-init on the "calls" var (`isAccessedBy(D, Init)` is true at the same time that `capturedByInit` is true).

There are some existing test cases that are similar and were passing: https://github.com/llvm/llvm-project/blob/0c0281130ed51fea06cf20a2db37c5bca257ad31/clang/test/CodeGenCXX/trivial-auto-var-init.cpp#L33

However, perhaps it's safer to just check `capturedByInit || isAccessedBy...`. WDYT? The capturedByInit could short-circuit in limited cases to avoid a second recursive pass over the `Init` too.

Also added another test case for stmt exprs (see below).

===

Looking at the implementation of `isCapturedBy` https://github.com/llvm/llvm-project/blob/0c0281130ed51fea06cf20a2db37c5bca257ad31/clang/lib/CodeGen/CGDecl.cpp#L1699 and `isAccessedBy`: https://github.com/llvm/llvm-project/blob/0c0281130ed51fea06cf20a2db37c5bca257ad31/clang/lib/CodeGen/CGDecl.cpp#L685

They are similar (both check BlockDecl captures), but do have some slight differences:
- isCapturedBy checks StmtExprs for VarDecl Inits that might recursively capture
- isAccessedBy checks DeclRefExprs while isCapturedBy does not

ilya-biryukov replied:
Ah, so `isAccessedBy` also detects all captured block variables. I have missed that when skimming through the code for the first time. Thanks for pointing that out.

If I'm not mistaken, `capturedByInit || ` does not actually change the semantics of the code, so not sure it adds any safety. But it does make sense as an optimization and that's also present in the other code for `isAccessedBy`, so including it seems like an overall win (performance, consistency).

ilya-biryukov reviewed (commented):
Looks from the C++ language rules and compiler frontend perspective.
The logic behinds the checks seems solid, there should not be any uncaught reads.

I am still unsure if we are doing the right thing for Objective C's `-fblocks`, left a comment asking about it.

jvoung committed:
Check capturedByInit and add test case for gnu statement expressions

In a few tests it appears that when capturedByInit is true,
isAccessedBy(Var, Init) is also true. capturedByInit also recurses
through the children of the Init expression. Just to be safe,
check capturedByInit || isAccessedBy -- and the capturedByInit
could help short-circuit in some cases and avoid repeating a
recursive traversal of children.

ilya-biryukov commented:
LGTM!

jvoung committed:
Merge branch 'main' into clang_less_init_scalars
